enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp

```diff
@@ -3861,6 +3861,18 @@ void TypeAnalyzer::visitIntrinsicInst(llvm::IntrinsicInst &I) {
         &I);
     return;
 
+  case Intrinsic::smax:
+  case Intrinsic::smin:
+  case Intrinsic::umax:
+  case Intrinsic::umin:
+    // No direction check as always valid
```

Comment:
I think this may be a bit too aggressive. 

E.g. I think here we would want:
  min(int, int) -> int
  min(ptr, ptr) -> ptr
 
 int return -> two args are int's
 ptr return -> two args are ptr's

Reply:
Thanks for the feedback, I have updated it. Let me know what you think. I am not sure if this part 

```
int return -> two args are int's
ptr return -> two args are ptr's
```

is covered by the changes.

Reply:
Not quite, you'd need to do something like

if (directions & up) {
  Auto retty = getAnalysis(I)[{-1}];
  if (retty == BaseType::Integer or Pointer) {
    Update analysis of op 0/1 with  that type tree
  }
}

if (directions & down) {
  Auto op0 = getAnalysis of operand 0, [{-1}]
  And op1
  If these are both int or both pointer
    update return analysis with corresponding Type tree
}

Reply:
Thanks for the help, I think understand a bit better

Reply:
I have updated the code as suggested, but it does not help my issue anymore. I have checked with my .ll file and the types are the following: (you see that they are never two pointers or two integers)

```
UP Unknown
DOWN Unknown Anything
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWNI nteger Anything
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Anything
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Integer Anything
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Integer
UP Unknown
DOWN Unknown Integer
```
Does it make sense to you if I add for DOWN that if any are integer then we propagate the integer?

Reply:
Does it suffice if we have integer  & (int or anything) -> int

And to test this we would copy the same two min/max except change one input argument to be 0.

Reply:
Unfortunately (int or anything) does not work for my case. Is it not fair to assume that if we have `InterType` on the first operand then necessarily we have another `InterType` on the second operand?

Reply:
I'd like to see the case where it fails with the int / anything first. Can you restore the int or anything functionality, and add the zero test. Then we can either add a second PR or second test for the too-conservative handling of that.

I worry that Int with unknown may be wrong/overly aggressive so want to be careful here.

Reply:
As an example, one side effect of int / unknown, means that the c = min(a : int, b) will always cause b to be an int. That may be valid, but I'm not sure. E.g. c will be set to an int since a is an int, then we can propagate up to b via c.

This may be problematic, for example if you have `min(16 : int, ptr-type)` which then will try to propagate int to the ptr and ptr to the int.

Reply:
Ok let me go back to anything and add the test!